Allow health checks to use any DI lifetime

rynowak opened the pull request:
This change allows registered IHealthCheck implementations to use any DI
lifetime. This is necessary for scenarios like using EF which requires a
scope.

The works by having the health check service create a scope for each
time it queries health checks. This scope does not overlap or share
state with other scopes (the request scope) so there is no crosstalk
between processing going on per-request in ASP.NET Core and the health
check operation.

rynowak commented on samples/HealthChecksSample/LivenessProbeStartup.cs:
```diff
@@ -53,11 +52,8 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
-                    "identity",
-                },
+                // Exclude all checks, just return a 200.
+                Predicate = (check) => false,
```
Changing this to an arbitrary predicate rather than an explicit 'filter by name' seems like a good usability improvement. Since the set of checks has to be resolved per-request a lot of motivations for providing 'filter by name' seemed to melt away.

rynowak commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/IHealthCheckService.cs:
```diff
@@ -1,26 +1,38 @@
-        /// The key maps to the <see cref="IHealthCheck.Name"/> property of the health check, and the value is the <see cref="IHealthCheck"/>
-        /// instance itself.
-        /// </remarks>
-        IReadOnlyDictionary<string, IHealthCheck> Checks { get; }
```
This has to go away now because they aren't singletons. The main use case we had for providing this was to support filtering. If someone wants to get the set of checks for an arbitrary reason they can create a scope and resolve them 😆

pranavkm commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthChecksBuilderAddCheckExtensions.cs:
```diff
@@ -88,5 +88,32 @@ public static IHealthChecksBuilder AddCheck(this IHealthChecksBuilder builder, I
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (typeof(T).IsAssignableFrom(typeof(IHealthCheck)))
```
`typeof(IHealthCheck).IsAssignableFrom(typeof(T))`

rynowak replied:
yeah going point. I think this was a bad copy-paste.

pranavkm commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthChecksBuilderAddCheckExtensions.cs:
```diff
@@ -88,5 +88,32 @@ public static IHealthChecksBuilder AddCheck(this IHealthChecksBuilder builder, I
+        /// provided implementation type <typeparamref name="T"/>. Using this method to register a health
+        /// check allows you to register a health check that depends on transient and scoped services.
+        /// </remarks>
+        public static IHealthChecksBuilder AddCheck<T>(this IHealthChecksBuilder builder) where T : class
```
Couldn't you do `where T : class, IHealthCheck`? Better than doing a runtime check

pranavkm commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs:
```diff
@@ -6,127 +6,103 @@
-            Checks = healthChecks.ToDictionary(c => c.Name, StringComparer.OrdinalIgnoreCase);
+                if (predicate != null)
+                {
+                    healthChecks = healthChecks.Where(predicate);
```
You could avoid the Linq call by doing this inside the `foreach`

rynowak replied:
sure

pranavkm commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs:
```diff
@@ -6,127 +6,103 @@
-                // If the health check does things like make Database queries using EF or backend HTTP calls,
-                // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
-                using (_logger.BeginScope(new HealthCheckLogScope(check.Name)))
+                if (!names.Add(check.Name))
```
This would print the same name multiple times if the check is registered more than twice. Think a `GroupBy` would be terse and more correct

rynowak replied:
sure good point.

davidfowl commented on src/Microsoft.AspNetCore.Diagnostics.HealthChecks/HealthCheckOptions.cs:
```diff
@@ -15,15 +15,19 @@ namespace Microsoft.AspNetCore.Diagnostics.HealthChecks
+        /// provide a function that filters the set of checks.
         /// </remarks>
-        public ISet<string> HealthCheckNames { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        public Func<IHealthCheck, bool> Predicate { get; set; }
```
Filter?

rynowak replied:
Eh, we considered it and thought this was less overloaded. If we use filter (or another term) other places in the framework then I want to change this.

davidfowl commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs:
```diff
@@ -6,127 +6,103 @@
+                        HealthCheckResult result;
+                        try
+                        {
+                            _logger.LogTrace("Running health check: {healthCheckName}", healthCheck.Name);
```
This isn't the logging pattern we use in components.

rynowak replied:
YEAH BRO

rynowak commented:
Updated this and cleaned up the logging. I logged an issue to consider the logging further, I don't think what's there is good enough but I don't want to drill into as part of fixing DI.

pranavkm commented on src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs:
```diff
@@ -6,127 +6,139 @@
-                        _logger.LogError(ex, "Health check '{healthCheckName}' threw an unexpected exception", check.Name);
-                        result = new HealthCheckResult(HealthCheckStatus.Failed, ex, ex.Message, data: null);
-                    }
+                public static readonly EventId HealthCheckBegin = new EventId(100, "HealthCheckBegin");
```
Nice
